source-mongodb/pull.go

```diff
@@ -279,477 +223,47 @@ func updateResourceStates(prevState captureState, bindings []bindingInfo) (captu
 		var sk = binding.stateKey
 		if resState, ok := prevState.Resources[sk]; ok {
 			newState.Resources[sk] = resState
+			trackedDatabases[binding.resource.Database] = struct{}{}
 		}
 	}
 
-	// Clear prior capture-wide state if there is no state for any currently included binding. This
-	// would happen if every binding had its backfill counter incremented, or all the bindings were
-	// removed from the spec. Notably, this removes any checkpointed StreamResumeToken, which may
-	// have become invalid.
-	if len(newState.Resources) == 0 {
-		if newState.OplogEmptyBeforeBackfill != nil || newState.StreamResumeToken != nil {
+	// Reset the database resume tokens if there are no active bindings for a given database.
+	for db, tok := range prevState.DatabaseResumeTokens {
+		if _, ok := trackedDatabases[db]; !ok {
 			log.WithFields(log.Fields{
-				"OplogEmptyBeforeBackfill": newState.OplogEmptyBeforeBackfill,
-				"StreamResumeToken":        newState.StreamResumeToken,
-			}).Info("clearing capture-wide state variables since all bindings have been removed or reset")
+				"database":   db,
+				"priorToken": tok,
+			}).Info("resetting change stream resume token for database")
+			continue
 		}
 
-		newState.OplogEmptyBeforeBackfill = nil
-		newState.StreamResumeToken = nil
+		newState.DatabaseResumeTokens[db] = tok
 	}
 
 	return newState, nil
 }
 
-func (s *captureState) wasChangeStreamEmptyBeforeBackfill() bool {
-	if s.OplogEmptyBeforeBackfill != nil {
-		return *s.OplogEmptyBeforeBackfill
-	}
-	return false
-}
-
-func (s *captureState) getBackfillStartTime() time.Time {
-	var start time.Time
-	for _, r := range s.Resources {
-		if start.IsZero() || start.After(r.Backfill.StartedAt) {
-			start = r.Backfill.StartedAt
-		}
-	}
-	return start
-}
-
 func (s *captureState) isBackfillComplete(bindings []bindingInfo) bool {
 	for _, b := range bindings {
 		if !s.Resources[b.stateKey].Backfill.Done {
 			return false
 		}
 	}
-	return len(s.Resources) > 0
+	return true
 }
 
 func (s *captureState) Validate() error {
 	return nil
 }
 
 type backfillState struct {
-	Done bool `json:"done,omitempty"`
-
-	StartedAt time.Time     `json:"started_at,omitempty"`
-	LastId    bson.RawValue `json:"last_id,omitempty"`
+	Done           bool           `json:"done,omitempty"`
+	LastId         *bson.RawValue `json:"last_id,omitempty"`
```

Comment:
Made this a pointer since `omitempty` doesn't really work with non-pointer structs